aws/lambda/log-classifier/src/engine.rs

```diff
@@ -44,3 +45,381 @@ pub fn evaluate_ruleset(ruleset: &RuleSet, log: &Log) -> Option<Match> {
+///
+/// # Arguments
+///
+/// * `ruleset` - A reference to the RuleSet to evaluate.
+/// * `log` - A reference to the Log to evaluate against.
+///
+/// # Returns
+///
+/// A vector of line numbers that match any rule in the ruleset.
+/// Note: Each rule in ruleset can only match 1 line which is the last line that matches the rule.
+pub fn evaluate_ruleset_all_lines(ruleset: &RuleSet, log: &Log) -> Vec<usize> {
+    let matches: Vec<Match> = ruleset
+        .rules
+        .par_iter()
+        .flat_map(|rule| evaluate_rule(rule, log))
+        .collect();
+    // combine captures into a single vector
+    let mut lines = Vec::new();
+    for m in matches {
+        lines.push(m.line_number);
+    }
+    lines
+}
+
+/// Merges line numbers into chunks based on specified size constraints.
+///
+/// # Arguments
+///
+/// * `log` - A reference to the Log structure containing the lines.
+/// * `line_numbers` - A vector of line numbers to be chunked.
+/// * `min_context_padding` - The minimum size of each chunk.
+/// * `max_chunk_size` - The maximum size of each chunk.
+///
+/// # Returns
+///
+/// A vector of tuples, where each tuple contains the start and end line numbers of a chunk.
+/// The chunks contain a contiguous set of lines around the specified line numbers. We extend
+/// the chunks by `min_context_padding` on both sides, up to the limits of the log which could
+/// lead to overlapping chunks.
+///
+/// # Panics
+///
+/// This function will panic if:
+/// * `min_context_padding` is greater than or equal to `max_chunk_size`.
+/// * Any line number is less than 1 or greater than the number of lines in the log.
+fn get_line_number_chunks(
+    log: &Log,
+    line_numbers: Vec<usize>,
+    min_context_padding: usize,
+    max_chunk_size: usize,
+) -> Vec<(usize, usize)> {
+    // throw user error if min_context_padding is greater than max_chunk_size
+    if min_context_padding * 2 >= max_chunk_size {
+        panic!("min_context_padding cannot be greater or equal to max_chunk_size");
+    }
+
+    // check if empty
+    if line_numbers.is_empty() {
+        return Vec::new();
+    }
+
+    // sort line numbers
+    let mut line_numbers = line_numbers;
+    line_numbers.sort();
+
+    // assert length of log is at least last line number and no negative line numbers
+    if *line_numbers.first().unwrap() < 1 || *line_numbers.last().unwrap() > log.lines.len() {
+        panic!("line numbers must be within the range of the log");
+    }
+
+    // merge line numbers into chunks
+    let mut merged_line_numbers = Vec::new();
+    // as we add min_context_padding to both sides
+    let effective_max_chunk_size = max_chunk_size - (min_context_padding * 2);
+    let mut chunk_start = 0;
+    let mut chunk_end = 0;
+    for line_number in line_numbers {
+        if chunk_start == 0 {
+            chunk_start = line_number;
+            chunk_end = line_number;
+        } else if line_number - chunk_start < effective_max_chunk_size {
```

Comment:
I believe the comparison should look like this:
```
line_number - chunk_start < effective_max_chunk_size + min_context_padding * 2
```

since afterwards you're trying to expand the chunk by `min_context_padding` from both directions.

Reply:
We pad to both sides so this is not true. Also effective_max_chunk_size = max_chunk_size - (min_context_padding * 2) which captures this, so we don't want to subtract padding twice

Reply:
Oh, yeah, you're right, I missed the definition of `effective_max_chunk_size`, thought you're comparing to just `max_chunk_size`.